torch/_subclasses/fake_tensor.py

```diff
@@ -821,40 +821,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         # is written to must be invalidated
         self.invalidate_written_to_constants(func, flat_arg_fake_tensors, args, kwargs)
 
-        from torch._decomp import decomposition_table
-
-        with self:
-            # Decomposes CompositeImplicitAutograd ops
-            r = func.decompose(*args, **kwargs)
-            if r is not NotImplemented:
-                return r
+        # If there's a Python meta, prefer that over the decomposition
+        from torch._decomp import meta_table as meta_table
 
-        # IDK: feels bad man, sym_numel on as_strided infinite loops otherwise
-        if has_symbolic_sizes and not self.cpp_meta_supports_symint(func):
-            from torch._decomp import meta_table as meta_table
+        if func not in meta_table:
```

Comment:
I had always have this question? 

Why FakeTensor need to explicitly call the functions in meta_table and decomp_tables? 
Why it cannot reply on the MetaTensor under the hood, to reach the meta DispatchKey, and then call into the meta functions? 

IIUC, such path is still executed for C++ meta function at line 875. 
In meta_registration.py, we have already registered a bunch of python meta functions to the C++ dispatcher. 
So invoking meta function from C++ dispatcher should still work ?

Reply:
So, to be clear, we are NOT explicitly calling `meta_table` anymore. We used to, but now we just use the meta table to try to figure out if there is a meta function that is likely to work, and if so, we'll go to the "classic" fake tensor algorithm, which is to run the meta function, and then try to figure out what the device is supposed to be afterwards. We're using the Python meta table as a proxy for "works with symbolic shapes", since Python metas automatically work with symbolic shapes. (And post your review, I readded the "cpp supports symints" as the other condition).

The question about decomp tables still stands, however. There seem to be two things going on here:

1. For the symbolic shapes case, I want to avoid hitting the C++ decomposition when there's a perfectly good Python one. But as you point out, we ought to have registered the Python decomp to the Meta key, so we'll hit it anyway. This is a good point, and I will test this.

2. Sometimes, it's better to decompose to fake tensor operations, than it is to decompose to meta tensor operations and fakeify post facto. @IvanKobzarev has been grappling with this in https://github.com/pytorch/pytorch/pull/88998 . Decomposing to fake tensor is preferable because some type promotion behavior is device sensitive, and so if you go all the way to meta you've lost important info that affects if something should pass checks or not (that being said, you can, and we in practice do, hack it up by using `device_hint` in meta registrations, so this ain't a perfect explanation). But there's also a kind of BE argument: if you have a decomposition, it ostensibly figures out device prop correctly intrinsically, so why go through the trouble of destroying the devices and then forcing fake tensor to figure it out again afterwards.

Now, what this PR does not do, is figure out when we should do (2). Empirically, this problem doesn't seem to happen with aten programs. Ivan only ran into it working with nvprims. I'm kind of thinking device_hint might be the right way to get decomps working correctly in these cases.

Reply:
Currently, on master, all `torch.ops.prims` go through the meta functions with fake tensors, not meta tensors, `in_kernel_invocation_manager` is not used:
https://github.com/pytorch/pytorch/blob/f20b3f2e5734b23a9e0a898196ddf77aa90323b8/torch/_subclasses/fake_tensor.py#L864-L866

Reply:
I'm going to land this PR as is, because it looks like it passes CI, and the more rigorous decomp versions don't. But I will keep running experiments to see if we can keep simplifying.